### DISCUSSION TITLE
Branches Difference

### DISCUSSION THREAD

**Discussion by Compr0mzd** [tobozo/WUD-Ducky#12] [Q&A]
Hey I was following a bit the project what are the difference between the RC1 branch and the non RC1 please . Thanks in advance keep up the good work !

**Comment by tobozo** [comment]
hey @Compr0mzd the branch versions of this project match the espressif arduino core versions they have been tested with.

I haven't been able to test with arduino core 2.0.3 yet but will create the branch soon, but *in theory* 2.0.3-RC1 branch should work with arduino core 2.0.3.

![image](https://user-images.githubusercontent.com/1893754/172348852-1ba63b1e-0508-47bc-be86-22a4e83d0815.png)

**Reply by Compr0mzd** [reply]
Ok thanks a lot !

**Reply by Compr0mzd** [reply]
Do you have any little tutorial step by step to make your repo work ?

**Reply by tobozo** [reply]
not sure this needs a tutorial, it's like any other Arduino project: load the ino file in the Arduino IDE; provided the library dependencies are installed, set the WUD in flash mode and hit the flash button.

did I miss a step ?

**Reply by Compr0mzd** [reply]
Nope but on the main page people might understand everything. I just got it working without trouble for now

**Reply by tobozo** [reply]
the main page is still on an older version of the ReadMe, the updated version has a list of library dependencies

I'm not good at writing tutorials (more of a data dumper) but your suggestion makes sense so I've activated the Wiki and disabled restrictions

![image](https://user-images.githubusercontent.com/1893754/172373457-2092ea88-4c4f-4e1d-ab95-109003d31c1c.png)

**Reply by Compr0mzd** [reply]
Great i'll probably make some content in there ! I saw on the other discussion that you included the wud-nutquacker.local domain is it working out of the box or we need to make some changes ? I'm getting disconnected from the Wifi when I navigate into the Webmenu is it a well known bug for you ?

**Reply by tobozo** [reply]
the mdns name follows the usb product name in the config, unless overridden from pins_arduino.h (as some adafruit boards do)

https://github.com/tobozo/WUD-Ducky/blob/arduino-esp32-2.0.3-RC1/USBDevices/USBConfig.h#L27

do you get a new device notification on your computer right after this occurs ? if so then it's a sign of a crash

the usual curlpit is an incomplete or corrupted spiffs partition, which may happen after switching boards from the arduino tools menu, or after flashing another sketch

**Reply by Compr0mzd** [reply]
I connect to thhe WIFI and when i load the WEB UI 
![image_2022-06-07_141850496](https://user-images.githubusercontent.com/83233183/172377049-b3b3b503-f327-4d04-b7bf-e8261f1ceb54.png)
Windows show me a new device connected
On the WEB UI i got only one time the result of the info page :
![image](https://user-images.githubusercontent.com/83233183/172377224-0b71f688-acfb-4a28-a1d0-4e2dafc54617.png)
When I change the Filesystem in the config the Wifi get disconnected. I don't see my SD card storage only the little FS volume :
![image](https://user-images.githubusercontent.com/83233183/172377464-e4cc4bf2-34fc-4235-ae35-ae4f9dcbd198.png)
I'm using the 2.0.3 (not RC1) core so it could be the problem.
Could it be my SD card or SD card library ? Formatted to FAT32 and tried NTFS as well no differences. 
Do you have any idea that could come in mind 

EDIT : 
my mDNS is http://esp32-wud.local/ you were right.
EDIT 2 : it seems that the logging function seems to make it crash. I had the time to connect to the Wifi and to disable the logging and when i did that I could see the files, change the file system, have the status in the info page + not getting disconnected

**Reply by tobozo** [reply]
> not getting disconnected

oh good, does it mean you can now access the system info page without crash?

I just did a test with 2.0.3, even re-applied the ESP32 sketch data upload over OTA but can't reproduce the crash.

Maybe double-check you're really using `LittleFS` as the partition type to upload?

![image](https://user-images.githubusercontent.com/1893754/172389852-af4bd030-5221-4fb4-a200-90ef1a2a482b.png)

With 2.0.3 the SD behaviour doesn't look good though, repeated errors have the volume mounted as readonly on my system.

![image](https://user-images.githubusercontent.com/1893754/172389456-76b49ef3-98b2-4acf-9419-38e258a66498.png)

I'm still not sure what's going on, within the Arduino core ESP32-S2 profile, all native rx/tx pins are mixed up, aliased or hidden behind macros only to support a tinyUSB integration that looks like a mess, so it's a real pain in the ass to debug on this platform :roll_eyes: 

I have already tried enabling the CDC interface but it makes the keyboard/mouse composite or the MSC inoperative.
USB-CDC is suboptimal for debug as it is unable to forward a crash dump since it relies on software to work :facepalm: 

I'll need some time to research whether this is caused by a tinyUSB glitch or recent changes with the SD driver, but both projects have a very high inertia so I wouldn't expect a solution for that very soon, better live with readonly SD for the meantime.

**Reply by Compr0mzd** [reply]
Yes i saw as well that the system is mounted on Read-Only . I tried to tweak things out without a luck. I got rid of the SD card to see if it was making it crash but no even without the SD card now it's crashs and i get disconnected from the SSID. How do you chose the option of the FS on Arduino ? Can't find the option you have or I don't know where is it located
EDIT : By doing http://esp32-wud.local/changefs?fs=sd  before the ES32 tries to load the files on the WebUI makes it not crash for me (I don't even have an SD card inside and still seems to say it worked fine) What a nightmare to debug x)

**Comment by tobozo** [comment]
so far when the partition is mounted I can read ~504 bytes before getting an i/o error, the write limitation is probably similar.

However I have flashed a simple SD_Test.ino example and the SD driver appears to operate correctly.
**Can you confirm such a basic example works for you too?** This may eventually expose a bad SD Card, or a bad contact on the PCB.

> How do you chose the option of the FS on Arduino

do you mean to upload LittleFS ? I have it in my tools menu.

https://github.com/lorol/arduino-esp32fs-plugin

![image](https://user-images.githubusercontent.com/1893754/172422687-04ddbaec-ec28-4adf-82c9-e52d28f6ba87.png)

**Reply by Compr0mzd** [reply]
Hi I'm gonna check it this morning ! I'll update

**Reply by Compr0mzd** [reply]
I installed the plugin and uploaded some data with LittleFS no problem except the error I have 
```
->Leaving...
->WARNING: ESP32-S2FH4 chip was placed into download mode using GPIO0.
->esptool.py can not exit the download mode over USB. To run the app, reset the chip manually.
->To suppress this note, set --after option to 'no_reset'.
LittleFS Upload failed!
```
After reuploading the WUD sketch i get finally to see the LittleFS online ! working properly 
![image](https://user-images.githubusercontent.com/83233183/172562668-92049a10-262d-4896-a923-33bf61db224e.png)
![image](https://user-images.githubusercontent.com/83233183/172562709-0cae62e0-7da3-48de-aa9a-1cc46e7c086d.png)
I have write access as well. I'm gonna try to investigate on the SD part now.

**Reply by Compr0mzd** [reply]
I cannot see the files present on the SD card that is formatted to exFAT. At least I can upload some scripts on the LittleFS

**Reply by tobozo** [reply]
> ->esptool.py can not exit the download mode over USB. To run the app, reset the chip manually.

although the IDE uses red colors that's not really an error, more of an incomplete implementation between esptool and arduino IDE :-)

you probably won't get that error when uploading LittleFS with OTA

some interesting facts about windows: it will cache the USB decriptors between insertions for faster loading, so if you had a previous error that ended up with the esp32-wud producing a messed up USB descriptor, chances are windows is still associating the corrupted version of that descriptor to the same PID/VID

this poisoned setting can be flushed with a **cold boot** (warn: some laptops use Fast Startup and this makes it worse)

I've also observed the ESP32-WUD acting differently when depending on when it was detected e.g. using a bootable USB port, or boot the laptop with the dongle plugged in so it is detected before the touchpad and other usb gadgets. 

This highlights the need to be able to dynamically change the device order and id assignation on the tinyUSB side to prevent class conflicts with the existing devices on the OS side. Given the inertia on the tinyUSB repo I can't imagine this happening any time soon.

Now back to the SD not visible from the WebUI:

I'm not sure about extFAT, but I'm using FAT32

```log
$> fdisk -l /dev/sdc
Disk /dev/sdc: 7,43 GiB, 7969177600 bytes, 15564800 sectors
Disk model: Mass-Storage    
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xeffe3401

Device     Boot Start     End Sectors Size Id Type
/dev/sdc1        2048 8388607 8386560   4G  c W95 FAT32 (LBA)
```

```log
$> fdisk -l /dev/sdc1
Disk /dev/sdc1: 3,102 GiB, 4293918720 bytes, 8386560 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x00000000
```

The sector size is 512, and I just found out the default [exFAT allocation unit sizes](https://www.partitionwizard.com/partitionmagic/allocation-unit-size-exfat.html) range from 4KB to 128K, so this may be another lead.

**Reply by Compr0mzd** [reply]
Yeah it didn't cause any problems I just had to restart the WUD and worked . Really interesting to know how Windows behave ! Thanks for the explanation. I formatted the card in FAT32 and it's working ! Since I used the 2.0.3 core I think we can confirm it worked. The only issue or mistery left to explain is to understand why before using the plugin you provided the LittleFS system seems either corrupted or non working properly (in R/O) thanks again for making this project possible without you I would just have a useless USB dongle ! 
I'm too bad in programming in the first place so I can't imagine it working for small chips with low level behavior
![image](https://user-images.githubusercontent.com/83233183/172588697-ba1d3297-dbbc-4189-a9a6-1c1523899994.png)
![image](https://user-images.githubusercontent.com/83233183/172588743-bb99b015-e12c-44ad-9b37-196db28bd7f9.png)
